sdk/wallet/wallet.go

```diff
@@ -0,0 +1,189 @@
+package wallet
```

Comment:
Let's refactor the wallet methods a bit. This may need some refinement, but I think it will be easier for consumers to understand if we only work on the seed phrase rather than the wallet seed.

```go
package wallet

import (
	"bytes"
	"crypto/ed25519"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"syscall/js"

	"go.sia.tech/core/consensus"
	"go.sia.tech/core/types"
	"go.sia.tech/coreutils/wallet"
	"go.sia.tech/web/sdk/marshal"
	"go.sia.tech/web/sdk/utils"
)

func encodeToBytes(v types.EncoderTo) []byte {
	var buf bytes.Buffer
	e := types.NewEncoder(&buf)
	v.EncodeTo(e)
	e.Flush()
	return buf.Bytes()
}

// GenerateSeedPhrase returns a new seed phrase.
func GenerateSeedPhrase(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	return map[string]any{
		"phrase": wallet.NewSeedPhrase(),
		"error":  nil,
	}
}

// GenerateKeyPair returns a new ed25519 key pair.
func GenerateKeyPair(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	pk := types.GeneratePrivateKey()
	return map[string]any{
		"privateKey": hex.EncodeToString(pk),
		"publicKey":  pk.PublicKey(),
	}
}

// KeyPairFromSeed returns the key pair corresponding to the given seed phrase and index.
func KeyPairFromSeed(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString, js.TypeNumber); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	var seed [32]byte
	if err := wallet.SeedFromPhrase(&seed, args[0].String()); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}
	priv := wallet.KeyFromSeed(&seed, uint64(args[1].Int()))
	return map[string]any{
		"privateKey": hex.EncodeToString(priv),
		"publicKey":  priv.PublicKey(),
	}
}

// StandardUnlockConditions returns the unlock conditions for a standard v1 unlockhash.
func StandardUnlockConditions(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	var pub types.PublicKey
	if pub.UnmarshalText([]byte(args[0].String())) != nil {
		return map[string]any{
			"error": "invalid public key",
		}
	}
	uc := types.StandardUnlockConditions(pub)
	obj, err := marshal.MarshalStruct(uc)
	if err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error marshaling unlock conditions: %s", err),
		}
	}
	return map[string]any{
		"unlockConditions": obj,
	}
}

// StandardUnlockHash returns the v1 unlockhash corresponding to the given public key.
func StandardUnlockHash(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	var pub types.PublicKey
	if pub.UnmarshalText([]byte(args[0].String())) != nil {
		return map[string]any{
			"error": "invalid public key",
		}
	}
	return map[string]any{
		"address": types.StandardUnlockHash(pub).String(),
	}
}

// AddressFromUnlockConditions returns the address corresponding to the given unlock conditions.
func AddressFromUnlockConditions(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	var uc types.UnlockConditions
	if err := json.Unmarshal([]byte(args[0].String()), &uc); err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error decoding unlock conditions: %s", err),
		}
	}

	return map[string]any{
		"address": uc.UnlockHash().String(),
	}
}

// AddressFromSpendPolicy returns the address of a spend policy.
func AddressFromSpendPolicy(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	var sp types.SpendPolicy
	if err := json.Unmarshal([]byte(args[0].String()), &sp); err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error decoding spend policy: %s", err),
		}
	}
	return map[string]any{
		"address": sp.Address().String(),
	}
}

// EncodeTransaction returns the binary encoding of a transaction.
func EncodeTransaction(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	jsonTxn := args[0].String()

	var txn types.Transaction

	if err := json.Unmarshal([]byte(jsonTxn), &txn); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	encoded := encodeToBytes(txn)
	return map[string]any{
		"encodedTransaction": encoded,
	}
}

// SignTransaction returns the signature of a transaction.
func SignTransaction(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString, js.TypeString, js.TypeString, js.TypeNumber, js.TypeString); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	jsonCs := args[0].String()
	jsonCn := args[1].String()
	jsonTxn := args[2].String()
	sigIndex := args[3].Int()

	buf, err := hex.DecodeString(args[4].String())
	if err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error decoding private key: %s", err),
		}
	} else if len(buf) != ed25519.PrivateKeySize {
		return map[string]any{
			"error": fmt.Sprintf("invalid private key length: %d", len(buf)),
		}
	}
	privateKey := types.PrivateKey(buf)

	var cs consensus.State
	if err := json.Unmarshal([]byte(jsonCs), &cs); err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error decoding consensus state: %s", err),
		}
	}

	var cn consensus.Network
	if err := json.Unmarshal([]byte(jsonCn), &cn); err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error decoding consensus network: %s", err),
		}
	}

	var txn types.Transaction
	if err := json.Unmarshal([]byte(jsonTxn), &txn); err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error decoding transaction: %s", err),
		}
	}
	cs.Network = &cn

	tsig := &txn.Signatures[sigIndex]
	var sigHash types.Hash256
	if tsig.CoveredFields.WholeTransaction {
		sigHash = cs.WholeSigHash(txn, tsig.ParentID, tsig.PublicKeyIndex, tsig.Timelock, tsig.CoveredFields.Signatures)
	} else {
		sigHash = cs.PartialSigHash(txn, tsig.CoveredFields)
	}
	return map[string]any{
		"signature": privateKey.SignHash(sigHash).String(),
	}
}

// TransactionID returns the ID of a transaction.
func TransactionID(this js.Value, args []js.Value) any {
	if err := utils.CheckArgs(args, js.TypeString); err != nil {
		return map[string]any{
			"error": err.Error(),
		}
	}

	var txn types.Transaction
	if err := json.Unmarshal([]byte(args[0].String()), &txn); err != nil {
		return map[string]any{
			"error": fmt.Sprintf("error decoding transaction: %s", err),
		}
	}

	return map[string]any{
		"id": txn.ID().String(),
	}
}
```

Creating a seed phrase and address:
```js
const seedPhrase = wallet.generateSeedPhrase(),
  { privateKey, publicKey } = wallet.keypairFromSeed(seedPhrase),
  address =  wallet.standardUnlockHash(publicKey);

const txn = buildTransaction(...),
  cs = getConsensusState(),
  cn = getConsensusNetwork();

txn.signatures[0].signature =  = wallet.signTransaction(cs, cn, txn, 0, privateKey);
```

instead of
```js
const seedPhrase = wallet.newSeedPhrase(),
  seed = wallet.seedFromPhrase(seedPhrase),
  { address } = wallet.publicKeyAndAddressFromSeed(seed, 0),
  privateKey = wallet.privateKeyFromSeed(seed, 0);

let txn = buildTransaction(...),
  cs = getConsensusState(),
  cn = getConsensusNetwork();


txn = wallet.signTransaction(cs, cn, txn, 0, privateKey);

Reply:
👍🏼  I can make these updated. Does this workflow move away from using the actual seed, in favor of seed phrase? The current walletd code operated primarily on the actual seed, this is what it stores in the cache, and also it uses for calculating the seedHash that gets checked when unlocking a wallet to send or generate addresses, eg:
```
const seedHash = blake2bHex(seed)
...
const { seed } = seedFromPhrase(mnemonic)
...
blake2bHex(seed) === seedHash || 'seed does not match'
```
Should we move to hashing the actual phrase?

Reply:
Yeah, hash the phrase itself